File: module19,decision tree and random forest/assign1.R

# import data
company <- read.csv("C:/Users/USER/Downloads/Company_Data.csv")

# sales is the target variable so we will be converting it into categorical form.
# categories are low=(<5), medium=(5 to 10), high=(10 to 15), very high= (>15))
sales_cat <- cut(company$Sales, breaks =c(0,5,10,15,20), labels = c("low","medium","high","very high"),right = FALSE)
sales_cat

any(is.na(company))

comp <- cbind(company,sales_cat)

# splitting data into training and testing

smp_size <- floor(0.75 * nrow(comp))

## set the seed to make your partition reproducible
set.seed(123)
train_ind <- sample(seq_len(nrow(comp)), size = smp_size)

train <- comp[train_ind, ]
test <- comp[-train_ind, ]

table(comp$sales_cat)

# Step 3: Training a model on the data
install.packages("C50")
library(C50)
sales_model <- C5.0(train[ ,2:11], train[ ,12])
plot(sales_model)

train_pred <- predict(sales_model, train[ ,2:11])
train_acc <- mean(train[ ,12]==train_pred)
train_acc # 93.66%

# test accuracy
test_pred <- predict(sales_model, test[ ,2:11])
test_acc <- mean(test[ ,12]==test_pred)
test_acc # 61%

# cross tabulation of predicted versus actual classes
install.packages("gmodels")
library(gmodels)
CrossTable(test[ ,12], test_pred, prop.chisq = FALSE, prop.c = FALSE, prop.r = FALSE,
           dnn = c('actual default', 'predicted default'))

File: module19,decision tree and random forest/assign1.py
# -*- coding: utf-8 -*-
"""
Created on Tue Jun  9 22:17:44 2020

@author: USER
"""

import pandas as pd
import numpy as np

company = pd.read_csv("C:/Users/USER/Downloads/Company_Data.csv")
company.head(15)

# we will convert target variable Sales into categorical form

sales_cat = pd.cut(company.Sales,bins=[0,5,10,15,20],labels=['low','medium','high','very high'])
company["sales"]= sales_cat

from sklearn.preprocessing import LabelEncoder
lb=LabelEncoder()
company["ShelveLoc"]=lb.fit_transform(company["ShelveLoc"])
company["Urban"]=lb.fit_transform(company["Urban"])
company["US"]=lb.fit_transform(company["US"])

company['sales'].unique()
company['sales'].value_counts()
colnames = list(company.columns)
type(company.columns)
predictors = colnames[1:11]
target = colnames[11]
target

# Splitting data into training and testing data set
from sklearn.model_selection import train_test_split
train,test = train_test_split(company,test_size = 0.2)

from sklearn.tree import DecisionTreeClassifier as DT

help(DT)
model = DT(criterion = 'entropy')
model.fit(train[predictors],train[target])

# Prediction on Train Data
preds = model.predict(train[predictors])
pd.crosstab(train[target],preds,rownames=['Actual'],colnames=['Predictions'])

np.mean(preds==train[target]) # 100%


# Prediction on Test Data
preds = model.predict(test[predictors])
pd.crosstab(test[target],preds,rownames=['Actual'],colnames=['Predictions'])

np.mean(preds==test[target]) # 60%













